### PR TITLE
(#4) Pagination of Results

### LINKED ISSUE #4: Story: Pagination of Results
# As a user, I would like to be able to page through search results from both the top and bottom of the page so that I can browse my search results.

** The updated pager should already be included in react-app-playground

> **ESTIMATE** 8

### Acceptance Criteria

```gherkin
  Scenario: View manual listing page results
    Given "trialListingPageType" is set to "Manual"
      And "pageTitle" is set to "Clinical Trials for Adult Metastatic Brain Tumors"
    Given the user navigates to "/?cfg=4"
     Then the page title is "Clinical Trials for Adult Metastatic Brain Tumors"
     Then the system displays 1 paragraph "Clinical trials are research studies that involve people. The clinical trials on this list are for breast cancer. All trials on the list are supported by NCI."
     Then the system displays 2 paragraph "NCI’s basic information about clinical trials explains the types and phases of trials and how they are carried out. Clinical trials look at new ways to prevent, detect, or treat disease. You may want to think about taking part in a clinical trial. Talk to your doctor for help in deciding if one is right for you."
      And the link "basic information about clinical trials" to "/about-cancer/treatment/clinical-trials/what-are-trials" appears on the page
      And each result displays the trial title as a link to the trial description page
      And each result displays the trial description below the link
      And each result displays "Location: " below the description
      And pager displays the following navigation options
      | pages  | 
      | 1      | 
      | 2      | 
      | 3      | 
      | 4      | 
      | 5      | 
      | Next > | 
  
  Scenario: View manual listing page metadata
    Given "trialListingPageType" is set to "Manual"
      And "pageTitle" is set to "Clinical Trials for Adult Metastatic Brain Tumors"
     When the user navigates to "/?cfg=4"
     Then the title tag should be "Clinical Trials for Adult Metastatic Brain Tumors - National Cancer Institute"
      And the page contains meta tags with the following properties
      | property       | content                                                      | 
      | og:title       | Clinical Trials for Adult Metastatic Brain Tumors            | 
      | og:url         | http://localhost:3000/                                       | 
      | og:description | Find clinical trials to treat adult metastatic brain tumors. | 
      And the page contains meta tags with the following names
      | name        | content                                                      | 
      | description | Find clinical trials to treat adult metastatic brain tumors. | 
      And there is a canonical link with the href "https://www.cancer.gov/"
  
  Scenario: User is able to navigate through pages on manual listings
    Given "trialListingPageType" is set to "Manual"
      And "pageTitle" is set to "Clinical Trials for Adult Metastatic Brain Tumors"
      And "itemsPerPage" is set to 10
    Given the user navigates to "/?cfg=4"
     Then the page title is "Clinical Trials for Adult Metastatic Brain Tumors"
     Then the system displays "Trials 1-10 of" "102"
      And pager displays the following navigation options
      | pages  | 
      | 1      | 
      | 2      | 
      | 3      | 
      | ...    | 
      | 11     | 
      | Next > | 
      And the page "1" is highlighted
     When user clicks on "Next >" button
     Then pager displays the following navigation options
      | pages      | 
      | < Previous | 
      | 1          | 
      | 2          | 
      | 3          | 
      | 4          | 
      | ...        | 
      | 11         | 
      | Next >     | 
      And the page "2" is highlighted
     When user clicks on "Next >" button
     Then pager displays the following navigation options
      | pages      | 
      | < Previous | 
      | 1          | 
      | 2          | 
      | 3          | 
      | 4          | 
      | 5          | 
      | ...        | 
      | 11         | 
      | Next >     | 
      And the page "3" is highlighted
     When user clicks on "Next >" button
     Then pager displays the following navigation options
      | pages      | 
      | < Previous | 
      | 1          | 
      | 2          | 
      | 3          | 
      | 4          | 
      | 5          | 
      | 6          | 
      | ...        | 
      | 11         | 
      | Next >     | 
      And the page "4" is highlighted
     When user clicks on "Next >" button
     Then pager displays the following navigation options
      | pages      | 
      | < Previous | 
      | 1          | 
      | ...        | 
      | 3          | 
      | 4          | 
      | 5          | 
      | 6          | 
      | 7          | 
      | ...        | 
      | 11         | 
      | Next >     | 
      And the page "5" is highlighted
     When user clicks on "11" button
     Then pager displays the following navigation options
      | pages      | 
      | < Previous | 
      | 1          | 
      | ...        | 
      | 9          | 
      | 10         | 
      | 11         | 
      And the page "11" is highlighted
     When user clicks on "< Previous" button
     Then pager displays the following navigation options
      | pages      | 
      | < Previous | 
      | 1          | 
      | ...        | 
      | 8          | 
      | 9          | 
      | 10         | 
      | 11         | 
      | Next >     | 
      And the page "10" is highlighted
     When user clicks on "< Previous" button
     Then pager displays the following navigation options
      | pages      | 
      | < Previous | 
      | 1          | 
      | ...        | 
      | 7          | 
      | 8          | 
      | 9          | 
      | 10         | 
      | 11         | 
      | Next >     | 
      And the page "9" is highlighted
     When user clicks on "< Previous" button
     Then pager displays the following navigation options
      | pages      | 
      | < Previous | 
      | 1          | 
      | ...        | 
      | 6          | 
      | 7          | 
      | 8          | 
      | 9          | 
      | 10         | 
      | 11         | 
      | Next >     | 
      And the page "8" is highlighted
     When user clicks on "< Previous" button
     Then pager displays the following navigation options
      | pages      | 
      | < Previous | 
      | 1          | 
      | ...        | 
      | 5          | 
      | 6          | 
      | 7          | 
      | 8          | 
      | 9          | 
      | ...        | 
      | 11         | 
      | Next >     | 
      And the page "7" is highlighted
  
  Scenario: View manual listing page with no pager (when total results is less than items per page )
    Given "trialListingPageType" is set to "Manual"
      And "pageTitle" is set to "Clinical Trials for Adult Metastatic Brain Tumors"
      And "itemsPerPage" is set to 20
    Given the user navigates to "/?cfg=4"
     Then the page title is "Clinical Trials for Adult Metastatic Brain Tumors"
      And pager is not displayed
```
```gherkin
Scenario: Mobile Pager Display
  Given user navigates to a listing page results URL
  Then the system returns the results page for cancer
    Then a horizontal rule is displayed below the "Trials 1-25 of XXX" 
    Then a pager is displayed
    Then a horizontal rule is displayed below the pager
    Then the result items are displayed 
    Then the option box to show 20 or 50 results is displayed
    Then a horizontal rule is displayed 
    Then the pager is displayed
    Then another horizontal rule is displayed below the pager
```

```gherkin
Scenario: Analytics for pager [TBD - @bryanpizzillo to add details here]
```

### Resources:
* Feature: JIRA Ticket #
* Mockups: *Here goes a URL to or the name of the mockup(s) in inVision*
* Web Page: *Here goes a URL to an existing web page*

### Notes

Example of the pager on prod: https://www.cancer.gov/about-cancer/treatment/clinical-trials/disease/myloproliferative-neoplasm/treatment

* > Here goes a quote from an email
* Here goes whatever useful information can exist…

---

### Prerequisites
- *required dependency for identifying the problem, scope and then solution for this issue*
- [ ] Short Spike Description - Issue #9999

### Tasks
- [x] Add pager to project - @seyilonge-nci 
- [x] Implement current pager - @KateMashkinaNIH 
- [x] Update pager - @seyilonge-nci 
- [x] Implement updated pager - @seyilonge-nci 

### Technical Notes
- More notes, or technical nuances/direction

### PR DESCRIPTION
**Pull request by seyilonge-nci**
Closes #4

* Added `Pager` to app
* Initial commit for pager update
* Pager component update, and implementation
* Added integration tests for pager
* Added mocks
* Changed pageUnit param in Manual component to be updated from state values
* Added unit tests for appendOrUpdateToQueryString
* Added unit tests for pager
* Added top/bottom pager section to `Manual`
* Updated @testing-library due to accessibility updated added to query DOM for aria attributes with 'hidden' value
* Manual unit test update
* Added pager to bottom of Manual component
* Updated unit and integration tests
* Removed unused 'listingType' from `ResultsList` prop
* Updated unit tests
* Removed conditional for onPageNavigationChange
* Added logic to hide jump menu when url param 'hideJumpMenu' has value of 'true'
* Fixed issues with 'getLocationInfoFromSites' util from previous rebase
* Updated unit tests
* String comparison for 'hideJumpMenu' value
* Added page neighbour implementation to pager
* Updated unit test and stylesheet
* Fixed bug with left ellipsis not showing based off current page and page neighbour
* Changed currentPageNeighbours to 2
* Updated pager integration tests
* Updated unit tests
* Added mocks
* Updated default page neighbour number, and max for pager
* Updated styling
* Added getPageOffset
* Updated integration tests
* Manual component update

### CODE REVIEW

**Comment by seyilonge-nci** [comment]
@VictoriaSunNIH @mworrest @blilianyu
Please review https://react-app-dev.cancer.gov/clinical-trials-listing-app/pr-30?cfg=4

File: src/components/molecules/pager/pager.scss [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+		}
+	}
+}
+
+/*@include break(medium) {
+	.pager__navigation,
+	.pager__container {
+		margin-left: auto;
+	}
+
```
Comment (KateMashkinaNIH) [inline src/components/molecules/pager/pager.scss:70]: If this not needed should probably be removed

**Review by KateMashkinaNIH — CHANGES_REQUESTED** [review]

File: src/components/molecules/pager/__tests__/pager.test.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+		// Previous pager item button should be hidden and have expected label
+		expect(
+			screen.getAllByRole('button', { hidden: true })[0]
+		).toHaveTextContent('Previous');
+
+		// Navigate to next page with next pager item button
+		fireEvent.click(screen.getByRole('button', { name: 'next page' }));
+
+		// Currently active page now should be page 2
+		/*expect(screen.getByRole('button', { name: 'page 2' })).toHaveAttribute(
```
[the remark this answers is not in the document]
Comment (seyilonge-nci) [reply src/components/molecules/pager/__tests__/pager.test.js:94]: Yes. I should have put a `todo` on that. Working on the unit test update.

File: src/components/molecules/pager/__tests__/pager.test.js [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+			container.querySelector('li.pager__ellipsis--left')
+		).toHaveTextContent('...');
+		expect(
+			container.querySelector('li.pager__ellipsis--right')
+		).toHaveTextContent('...');
+
+		// Click on last page number and confirm next button is hidden
+		fireEvent.click(screen.getByRole('button', { name: 'page 11' }));
+		/*expect(screen.getByRole('button', { name: 'page 11' })).toHaveAttribute(
+			'class',
```
[the remark this answers is not in the document]
Comment (seyilonge-nci) [reply src/components/molecules/pager/__tests__/pager.test.js:127]: Same as above

File: src/components/molecules/pager/__tests__/pager.test.js [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+		);*/
+
+		// Confirm onPageNavigationChange was called
+		const { onPageNavigationChange } = defaultProps;
+		expect(onPageNavigationChange).toHaveBeenCalled();
+
+		// Click on previous and confirm currently active page is 10
+		fireEvent.click(screen.getByRole('button', { name: 'previous page' }));
+		/*expect(screen.getByRole('button', { name: 'page 10' })).toHaveAttribute(
+			'class',
```
[the remark this answers is not in the document]
Comment (seyilonge-nci) [reply src/components/molecules/pager/__tests__/pager.test.js:138]: Yup, same as above.

**Comment by alinai** [comment]
Pager on IE:
image

**Comment by alinai** [comment]
The pager is not consistent with the AC and SWS when we are on page 4 (click Next when on Page 3 to go to Page 4)

image

image

image

**Comment by alinai** [comment]
The pager is not consistent with AC and Prod DTLP at mobile breakpoint.

image

**Comment by alinai** [comment]
"Trials 1-25 of XXX" should not appear at the bottom of the page. Horizontal rule is not showing up below the last result on the page. (See AC)

image

image

File: src/components/molecules/pager/__tests__/pager.test.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (seyilonge-nci) [reply src/components/molecules/pager/__tests__/pager.test.js]: Done

File: src/components/molecules/pager/__tests__/pager.test.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (seyilonge-nci) [reply src/components/molecules/pager/__tests__/pager.test.js:144]: Done

File: src/components/molecules/pager/__tests__/pager.test.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (seyilonge-nci) [reply src/components/molecules/pager/__tests__/pager.test.js:159]: Done

File: src/components/molecules/pager/pager.scss (hunk 1 above)
Comment (seyilonge-nci) [reply src/components/molecules/pager/pager.scss:70]: Updated

**Comment by VictoriaSunNIH** [comment]
@seyilonge-nci - I'm looking at mobile and seeing that "next" goes off the page (breaks the width of the page)

**Comment by alinai** [comment]
On Prod, at 1024px (tablet breakpoint) both the pager is aligned to the left. We are not seeing it happening on react.

**Comment by alinai** [comment]
Not seeing the horizontal rule below the last result at desktop and tablet breakpoint.

File: src/components/molecules/pager/pager.jsx [hunk 5]
```diff
@@ -0,0 +1,214 @@
+import PropTypes from 'prop-types';
+import React from 'react';
+
+import './pager.scss';
```
Comment (arcepaul) [inline src/components/molecules/pager/pager.jsx:4]: ack, sorry I didn't see this sooner.  Can we move the styling to app.scss?  Thinking about this as a component in a component library and styling wouldn't be married to the component

Comment (seyilonge-nci) [reply src/components/molecules/pager/pager.jsx]: Done!

**Review by arcepaul — APPROVED** [review]

**Pull request merged**